### PR TITLE
add service node

### PR DESCRIPTION
**Pull request by andre-merzky**
this PR adds a "services" key to the pilot description and resource
description.  Any commands listed there will be run by the agent on
a separate service node (which is then not usable for tasks anymore)

@mturilli : what do you think?

### CODE REVIEW

**Review by mturilli — COMMENTED** [review]
Looks a good starting point. I think we should probably assume that a service will need most of the time some network-related facility to be usable, like an IP or endpoint. Should we think about specific facilities to expose/make available that information?

File: src/radical/pilot/agent/agent_0.py [hunk 1]
```diff
@@ -322,6 +325,121 @@ def _write_sa_configs(self):
...
+            name    = self._cfg['agent_launch_method'],
+            cfg     = self._cfg,
+            session = self._session)
+
+        node = self._cfg['rm_info']['service_node']
+        bs_name = "%s/bootstrap_2.sh" % (self._pwd)
+        ls_name = "%s/services.sh"    % self._pwd
+        threads = self._cfg['rm_info']['cores_per_node']
+        slots   = {
+                    'cpu_processes'    : 1,
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/agent/agent_0.py:352]: Agree.

File: src/radical/pilot/agent/agent_0.py [hunk 2]
```diff
@@ -322,6 +325,121 @@ def _write_sa_configs(self):
...
+            ls.write('exec %s\n\n' % cmd)
+            st = os.stat(ls_name)
+            os.chmod(ls_name, st.st_mode | stat.S_IEXEC)
+
+        if hop : cmdline = hop
+        else   : cmdline = ls_name
+
+        # ------------------------------------------------------------------
+        class _S(mp.Process):
+
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/agent/agent_0.py:403]: :-P  Fixed.

File: src/radical/pilot/agent/bootstrap_0.sh [hunk 3]
```diff
@@ -1804,9 +1816,16 @@ export RADICAL_PILOT_NTPHOST=$RADICAL_PILOT_NTPHOST
...
 $PREBOOTSTRAP2_EXPANDED
 
-# start agent, forward arguments
+# start services and agent, forward arguments
 # NOTE: exec only makes sense in the last line of the script
-exec $AGENT_CMD "\$1" 1>"\$1.out" 2>"\$1.err"
+if test "\$1" = 'services'
+then
+    # start the services script
+    exec ./services 1> services.out 2> services.err
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/agent/bootstrap_0.sh:1824]: Maybe we need that eventually - let's address that when we have a use case?

**Comment by andre-merzky** [comment]
@mturilli : this is good for another review I think.

**Comment by andre-merzky** [comment]
> I think we should probably assume that a service will need most of the time some network-related facility to be usable, like an IP or endpoint. Should we think about specific facilities to expose/make available that information?

What kind of information do you have in mind?

**Comment by andre-merzky** [comment]
ping @mturilli : reade for another iteration.

**Review by mturilli — APPROVED** [review]
LGTM, thanks!

**Pull request merged**